tests/unit/data/data_pipeline/test_dynamic_bucket.py

```diff
@@ -0,0 +1,230 @@
+
+        pipeline = (
+            read_sequence(seq).dynamic_bucket(threshold, cost_fn, nb_min=2).and_return()
+        )
+
+        for _ in range(2):
+            it = iter(pipeline)
+
+            d = None
+            for i in range(1, 11, 2):
+                d = next(it)
+                assert d == [i, i + 1]
+
+            with pytest.raises(StopIteration):
+                next(it)
+
+            pipeline.reset()
+
+    def test_op_works_with_max_set(self) -> None:
+        seq = [0, 0, 0, 0, 1, 2, 3]
+
+        threshold = 3
+        cost_fn = lambda x: x
+
+        pipeline = (
+            read_sequence(seq).dynamic_bucket(threshold, cost_fn, nb_max=2).and_return()
+        )
+
+        for _ in range(2):
+            it = iter(pipeline)
+
+            assert next(it) == [0, 0]
+            assert next(it) == [0, 0]
+            assert next(it) == [1, 2]
+            assert next(it) == [3]
+
+            with pytest.raises(StopIteration):
+                next(it)
+
+            pipeline.reset()
+
+    def test_op_works_with_min_and_max_set(self) -> None:
+        seq = [0, 0, 0, 0, 1, 2, 3, 4]
+
+        threshold = 3
+        cost_fn = lambda x: x
+
+        pipeline = (
+            read_sequence(seq)
+            .dynamic_bucket(threshold, cost_fn, nb_min=2, nb_max=2)
+            .and_return()
+        )
+
+        for _ in range(2):
+            it = iter(pipeline)
+
+            assert next(it) == [0, 0]
+            assert next(it) == [0, 0]
+            assert next(it) == [1, 2]
+            assert next(it) == [3, 4]
+
+            with pytest.raises(StopIteration):
+                next(it)
+
+            pipeline.reset()
+
+    @pytest.mark.parametrize("drop", [False, True])
+    def test_op_works_with_min_and_drop_set(self, drop: bool) -> None:
+        seq = [1, 2, 4]
+
+        threshold = 3
+        cost_fn = lambda x: x
+
+        pipeline = (
+            read_sequence(seq)
+            .dynamic_bucket(threshold, cost_fn, nb_min=2, drop_remainder=drop)
+            .and_return()
+        )
+
+        for _ in range(2):
+            it = iter(pipeline)
+
+            assert next(it) == [1, 2]
+
+            if not drop:
+                assert next(it) == [4]
+
+            with pytest.raises(StopIteration):
+                next(it)
+
+            pipeline.reset()
+
+    def test_op_raises_error_when_threshold_is_nonpositive(self) -> None:
+        with pytest.raises(
+            ValueError, match=r"^`threshold` must be greater than zero\.$"
+        ):
+            read_sequence(list(range(100))).dynamic_bucket(0, lambda x: 1).and_return()
+
+        with pytest.raises(
+            ValueError, match=r"^`threshold` must be greater than zero\.$"
+        ):
+            read_sequence(list(range(100))).dynamic_bucket(-1, lambda x: 1).and_return()
+
+    def test_op_saves_and_restores_its_state(self) -> None:
+        seq = list(range(1, 7))
+
+        threshold = 2
+        cost_fn = lambda x: x
```

Comment:
let's take a least one example with different cost function (like x**2 for instance)